### PR TITLE
feature: support binance futures

### PR DESCRIPTION
**Pull request by austin362667**

### CODE REVIEW

File: pkg/exchange/binance/stream.go [hunk 1]
```diff
@@ -61,7 +62,9 @@ type Stream struct {
 	types.FuturesSettings
 	types.StandardStream
 
-	Client   *binance.Client
+	Client        *binance.Client
+	futuresClient *futures.Client
```
Comment (c9s) [inline pkg/exchange/binance/stream.go:66]: we can add this to the main branch first.

File: pkg/exchange/binance/stream.go [hunk 2]
```diff
@@ -76,13 +79,19 @@ type Stream struct {
 	kLineClosedEventCallbacks []func(e *KLineEvent)
 
 	markPriceUpdateEventCallbacks []func(e *MarkPriceUpdateEvent)
-	continuousKLineEventCallbacks []func(e *ContinuousKLineEvent)
+
+	continuousKLineEventCallbacks       []func(e *ContinuousKLineEvent)
+	continuousKLineClosedEventCallbacks []func(e *ContinuousKLineEvent)
```
Comment (c9s) [inline pkg/exchange/binance/stream.go:84]: we can add this to the main branch first.

File: pkg/exchange/binance/stream.go [hunk 3]
```diff
@@ -76,13 +79,19 @@ type Stream struct {
...
 
 	balanceUpdateEventCallbacks           []func(event *BalanceUpdateEvent)
 	outboundAccountInfoEventCallbacks     []func(event *OutboundAccountInfoEvent)
 	outboundAccountPositionEventCallbacks []func(event *OutboundAccountPositionEvent)
 	executionReportEventCallbacks         []func(event *ExecutionReportEvent)
 
+	accountUpdateEventCallbacks       []func(e *AccountUpdateEvent)
+	accountConfigUpdateEventCallbacks []func(e *AccountConfigUpdateEvent)
+	orderTradeUpdateEventCallbacks    []func(e *OrderTradeUpdateEvent)
```
Comment (c9s) [inline pkg/exchange/binance/stream.go:93]: we can add this to the main branch first.

File: pkg/exchange/binance/stream.go [hunk 4]
```diff
@@ -207,6 +216,174 @@ func NewStream(client *binance.Client) *Stream {
...
 
+	stream.OnContinuousKLineEvent(func(e *ContinuousKLineEvent) {
+		kline := e.KLine.KLine()
+		if e.KLine.Closed {
+			stream.EmitContinuousKLineClosedEvent(e)
+			stream.EmitKLineClosed(kline)
+		} else {
+			stream.EmitKLine(kline)
+		}
+	})
```
Comment (c9s) [inline pkg/exchange/binance/stream.go:227]: we can add this to the main branch first.

File: pkg/exchange/binance/stream.go [hunk 5]
```diff
@@ -207,6 +216,174 @@ func NewStream(client *binance.Client) *Stream {
...
+
+		if err != nil {
+			log.WithError(err).Error("subscribe error")
+		}
+	})
+
+	return stream
+}
+
+func NewFuturesStream(client *futures.Client) *Stream {
```
Comment (austin362667) [inline pkg/exchange/binance/stream.go:261]: merge this function into NewFuturesStream(), like this,
```
stream := NewStream(e.Client, e.futuresClient)
stream.MarginSettings = e.MarginSettings
stream.FuturesSettings = e.FuturesSettings
```

File: pkg/exchange/binance/convert.go [hunk 6]
```diff
@@ -270,6 +519,20 @@ func toGlobalSideType(side binance.SideType) types.SideType {
 	}
 }
 
+func toGlobalFuturesSideType(side futures.SideType) types.SideType {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:522]: in!

File: pkg/exchange/binance/convert.go [hunk 7]
```diff
@@ -292,6 +555,27 @@ func toGlobalOrderType(orderType binance.OrderType) types.OrderType {
 	}
 }
 
+func toGlobalFuturesOrderType(orderType futures.OrderType) types.OrderType {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:558]: done!

File: pkg/exchange/binance/convert.go [hunk 8]
```diff
@@ -313,10 +597,31 @@ func toGlobalOrderStatus(orderStatus binance.OrderStatusType) types.OrderStatus
 	return types.OrderStatus(orderStatus)
 }
 
+func toGlobalFuturesOrderStatus(orderStatus futures.OrderStatusType) types.OrderStatus {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:600]: done!

File: pkg/exchange/binance/convert.go [hunk 9]
```diff
@@ -313,10 +597,31 @@ func toGlobalOrderStatus(orderStatus binance.OrderStatusType) types.OrderStatus
...
+	}
+
+	return types.OrderStatus(orderStatus)
+}
+
 // ConvertTrades converts the binance v3 trade into the global trade type
 func ConvertTrades(remoteTrades []*binance.TradeV3) (trades []types.Trade, err error) {
 	for _, t := range remoteTrades {
-		trade, err := ToGlobalTrade(*t, false)
+		trade, err := toGlobalTrade(*t, false)
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:624]: done!

File: pkg/exchange/binance/convert.go [hunk 10]
```diff
@@ -163,6 +299,28 @@ func toLocalOrderType(orderType types.OrderType) (binance.OrderType, error) {
 	return "", fmt.Errorf("can not convert to local order, order type %s not supported", orderType)
 }
 
+func toLocalFuturesOrderType(orderType types.OrderType) (futures.OrderType, error) {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:302]: done!

File: pkg/exchange/binance/convert.go [hunk 11]
```diff
@@ -176,6 +334,19 @@ func toGlobalOrders(binanceOrders []*binance.Order) (orders []types.Order, err e
 	return orders, err
 }
 
+func toGlobalFuturesOrders(futuresOrders []*futures.Order) (orders []types.Order, err error) {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:337]: done!

File: pkg/exchange/binance/convert.go [hunk 12]
```diff
@@ -199,11 +370,36 @@ func toGlobalOrder(binanceOrder *binance.Order, isMargin bool) (*types.Order, er
 	}, nil
 }
 
+func toGlobalFuturesOrder(futuresOrder *futures.Order, isMargin bool) (*types.Order, error) {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:373]: done!

File: pkg/exchange/binance/convert.go [hunk 13]
```diff
@@ -199,11 +370,36 @@ func toGlobalOrder(binanceOrder *binance.Order, isMargin bool) (*types.Order, er
...
+		// IsIsolated:       futuresOrder.IsIsolated,
+	}, nil
+}
+
 func millisecondTime(t int64) time.Time {
 	return time.Unix(0, t*int64(time.Millisecond))
 }
 
-func ToGlobalTrade(t binance.TradeV3, isMargin bool) (*types.Trade, error) {
+func toGlobalTrade(t binance.TradeV3, isMargin bool) (*types.Trade, error) {
```
Comment (austin362667) [inline pkg/exchange/binance/convert.go:402]: done!

File: pkg/exchange/binance/exchange.go [hunk 14]
```diff
@@ -46,20 +46,38 @@ type Exchange struct {
 	types.MarginSettings
 	types.FuturesSettings
 
-	key, secret string
-	Client      *binance.Client
+	key, secret   string
+	Client        *binance.Client // Spot & Margin
+	futuresClient *futures.Client // USDT-M Futures
```
Comment (austin362667) [inline pkg/exchange/binance/exchange.go:51]: done!

File: pkg/exchange/binance/exchange.go [hunk 15]
```diff
@@ -46,20 +46,38 @@ type Exchange struct {
...
 	var client = binance.NewClient(key, secret)
 	client.HTTPClient = &http.Client{Timeout: 15 * time.Second}
-
 	_, _ = client.NewSetServerTimeService().Do(context.Background())
-	return &Exchange{
-		key:    key,
-		secret: secret,
 
-		Client: client,
+	var futuresClient = binance.NewFuturesClient(key, secret)
```
Comment (austin362667) [inline pkg/exchange/binance/exchange.go:60]: done!

File: pkg/exchange/binance/exchange.go [hunk 16]
```diff
@@ -152,8 +293,18 @@ func (e *Exchange) QueryAveragePrice(ctx context.Context, symbol string) (float6
 }
 
 func (e *Exchange) NewStream() types.Stream {
+
+	if e.IsMargin {
+		stream := NewStream(e.Client)
+		stream.MarginSettings = e.MarginSettings
+		return stream
+	} else if e.IsFutures {
```
Comment (austin362667) [inline pkg/exchange/binance/exchange.go:301]: refactored!

File: pkg/exchange/binance/exchange.go [hunk 17]
```diff
@@ -582,6 +874,95 @@ func (e *Exchange) submitMarginOrder(ctx context.Context, order types.SubmitOrde
 	return createdOrder, err
 }
 
+func (e *Exchange) submitFuturesOrder(ctx context.Context, order types.SubmitOrder) (*types.Order, error) {
```
Comment (austin362667) [inline pkg/exchange/binance/exchange.go:877]: done!

File: pkg/exchange/binance/exchange.go [hunk 18]
```diff
@@ -700,13 +1081,15 @@ func (e *Exchange) submitSpotOrder(ctx context.Context, order types.SubmitOrder)
...
 	for _, order := range orders {
-		if err := orderLimiter.Wait(ctx) ; err != nil {
+		if err := orderLimiter.Wait(ctx); err != nil {
 			log.WithError(err).Errorf("order rate limiter wait error")
 		}
 
 		var createdOrder *types.Order
 		if e.IsMargin {
 			createdOrder, err = e.submitMarginOrder(ctx, order)
+		} else if e.IsFutures {
```
Comment (austin362667) [inline pkg/exchange/binance/exchange.go:1091]: done!

File: pkg/exchange/binance/stream.go [hunk 19]
```diff
@@ -204,6 +216,174 @@ func NewStream(client *binance.Client) *Stream {
...
+			snapshot[balance.Asset] = types.Balance{
+				Currency:  balance.Asset,
+				Available: balance.Free,
+				Locked:    balance.Locked,
+			}
+		}
+		stream.EmitBalanceSnapshot(snapshot)
+	})
+
+	stream.OnAccountUpdateEvent(func(e *AccountUpdateEvent) {
```
Comment (austin362667) [inline pkg/exchange/binance/stream.go:281]: WIP

File: pkg/exchange/binance/stream.go [hunk 20]
```diff
@@ -204,6 +216,174 @@ func NewStream(client *binance.Client) *Stream {
...
+		// TODO
+	})
```
Comment (austin362667) [inline pkg/exchange/binance/stream.go:283]: WIP

File: pkg/exchange/binance/stream.go [hunk 21]
```diff
@@ -204,6 +216,174 @@ func NewStream(client *binance.Client) *Stream {
...
+
+	stream.OnAccountConfigUpdateEvent(func(e *AccountConfigUpdateEvent) {
+		snapshot := types.PositionMap{}
+		snapshot[e.AccountConfig.Symbol] = types.Position{
+			Symbol:   e.AccountConfig.Symbol,
+			Leverage: e.AccountConfig.Leverage,
+		}
+
+		stream.EmitPositionSnapshot(snapshot)
+	})
```
Comment (austin362667) [inline pkg/exchange/binance/stream.go:293]: should not modify position leverage.
update account config leverage only.

**Pull request closed without merging**